### PR TITLE
Support DOTNET_EnableDiagnostics in dd-dotnet

### PR DESCRIPTION
**Pull request by kevingosse**
## Summary of changes

The behavior of `DOTNET_EnableDiagnostics` has changed in .NET 8, it can now disable profiling: https://learn.microsoft.com/en-us/dotnet/core/tools/dotnet-environment-variables#dotnet_enablediagnostics
`DOTNET_EnableDiagnostics_Profiler` will have a similar effect: https://learn.microsoft.com/en-us/dotnet/core/tools/dotnet-environment-variables#dotnet_enablediagnostics_profiler

This PR makes two changes:
 - `dd-dotnet check` checks that those environment variable aren't set to 0
 - `dd-dotnet run` and `dd-trace run` explicitly sets them to 1 to override preventively override the ambient value
 
## Test coverage

Added an integration test to verify that `dd-dotnet check` detects those environment variables.

### CODE REVIEW

**Comment by andrewlock** [comment]
## Execution-Time Benchmarks Report :stopwatch:

Execution-time results for samples comparing the following branches/commits:
- [This PR (5208)](https://github.com/DataDog/dd-trace-dotnet/tree/5eb2070fd337a307827747cd3f104d1096de16df)
- [master](https://github.com/DataDog/dd-trace-dotnet/tree/0ebc75bd80b322ab3f1f0599c58db78ff69bff1d)

Execution-time benchmarks measure the whole time it takes to execute a program. And are intended to measure the one-off costs. Cases where the execution time results for the PR are worse than latest master results are shown in **red**. The following thresholds were used for comparing the execution times:
* Welch test with statistical test for significance of **5%**
* Only results indicating a difference greater than **5%** and **5 ms** are considered.


Note that these results are based on a _single_ point-in-time result for each branch. For full results, see the [dashboard](https://ddstaging.datadoghq.com/dashboard/4qn-6fi-54p/apm-net-execution-time-benchmarks).

Graphs show the p99 interval based on the mean and StdDev of the test run, as well as the mean value of the run (shown as a diamond below the graph).
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET Framework 4.6.2) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (5208) - mean (76ms)  : 67, 85
     .   : milestone, 76,
    master - mean (74ms)  : 66, 82
     .   : milestone, 74,

    section CallTarget+Inlining+NGEN
    This PR (5208) - mean (986ms)  : 963, 1008
     .   : milestone, 986,
    master - mean (985ms)  : 963, 1007
     .   : milestone, 985,

```
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET Core 3.1) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (5208) - mean (112ms)  : 108, 116
     .   : milestone, 112,
    master - mean (110ms)  : 107, 112
     .   : milestone, 110,

    section CallTarget+Inlining+NGEN
    This PR (5208) - mean (717ms)  : 693, 742
     .   : milestone, 717,
    master - mean (718ms)  : 696, 740
     .   : milestone, 718,

```
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET 6) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (5208) - mean (95ms)  : 91, 99
     .   : milestone, 95,
    master - mean (97ms)  : 93, 100
     .   : milestone, 97,

    section CallTarget+Inlining+NGEN
    This PR (5208) - mean (673ms)  : 650, 695
     .   : milestone, 673,
    master - mean (679ms)  : 647, 710
     .   : milestone, 679,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET Framework 4.6.2) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (5208) - mean (186ms)  : 182, 190
     .   : milestone, 186,
    master - mean (186ms)  : 180, 192
     .   : milestone, 186,

    section CallTarget+Inlining+NGEN
    This PR (5208) - mean (1,058ms)  : 1028, 1087
     .   : milestone, 1058,
    master - mean (1,052ms)  : 1027, 1078
     .   : milestone, 1052,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET Core 3.1) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (5208) - mean (269ms)  : 262, 276
     .   : milestone, 269,
    master - mean (270ms)  : 265, 275
     .   : milestone, 270,

    section CallTarget+Inlining+NGEN
    This PR (5208) - mean (1,057ms)  : 1034, 1081
     .   : milestone, 1057,
    master - mean (1,059ms)  : 1038, 1081
     .   : milestone, 1059,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET 6) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (5208) - mean (257ms)  : 252, 263
     .   : milestone, 257,
    master - mean (259ms)  : 254, 264
     .   : milestone, 259,

    section CallTarget+Inlining+NGEN
    This PR (5208) - mean (993ms)  : 969, 1017
     .   : milestone, 993,
    master - mean (992ms)  : 966, 1018
     .   : milestone, 992,

```

File: tracer/src/Datadog.Trace.Tools.dd_dotnet/Checks/ProcessBasicCheck.cs [hunk 1]
```diff
@@ -643,53 +694,22 @@ private static bool GetLocalMachineSubKeyVersion(string uninstallKey, string dat
...
-            }
-            else
-            {
-                Utils.WriteError(UnsupportedLinuxArchitecture(osArchitecture.ToString()));
-                return;
-            }
+            bool result = true;
 
-            try
+            foreach (var key in variablesToCheck)
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace.Tools.dd_dotnet/Checks/ProcessBasicCheck.cs:703]: Purely out of interest, can we use `[]` if we use it directly in a foreach?

```suggestion
            foreach (var key in ["DOTNET_EnableDiagnostics", "COMPlus_EnableDiagnostics", "DOTNET_EnableDiagnostics_Profiler", "COMPlus_EnableDiagnostics_Profiler"])
```

File: tracer/src/Datadog.Trace.Tools.dd_dotnet/Checks/Resources.cs [hunk 2]
```diff
@@ -57,6 +57,8 @@ internal static class Resources
 
         public static void ResetChecks() => _checkNumber = 1;
 
+        public static string EnableDiagnosticsSet(string key) => $"The environment variable {key} is set to 0, which disables profiling.";
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace.Tools.dd_dotnet/Checks/Resources.cs:60]: `COMPlus_EnableDiagnostics=0` isn't _actually_ a problem unless we're running under .NET 8 🤔 Do we care? Should we mention that it's ok if you're running <.NET 8?

**Review by andrewlock — APPROVED** [review]
LGTM

File: tracer/src/Datadog.Trace.Tools.dd_dotnet/Checks/Resources.cs (hunk 2 above)
Comment (kevingosse) [reply tracer/src/Datadog.Trace.Tools.dd_dotnet/Checks/Resources.cs:60]: I think it brings nothing and just increase the risks of false-negatives 😛

File: tracer/src/Datadog.Trace.Tools.dd_dotnet/Checks/ProcessBasicCheck.cs (hunk 1 above)
Comment (kevingosse) [reply tracer/src/Datadog.Trace.Tools.dd_dotnet/Checks/ProcessBasicCheck.cs:703]: Since it uses target-typing, nope. Doesn't compile :(

**Comment by andrewlock** [comment]
## Throughput/Crank Report:zap:

Throughput results for AspNetCoreSimpleController comparing the following branches/commits:
- [This PR (5208)](https://github.com/DataDog/dd-trace-dotnet/tree/5eb2070fd337a307827747cd3f104d1096de16df)
- [master](https://github.com/DataDog/dd-trace-dotnet/tree/f03ab8db35c818efcd9b84086075b82e741578ea)
- [benchmarks/2.9.0](https://github.com/DataDog/dd-trace-dotnet/tree/89ae3f4ee7d85b824badda5873166707c9e560b4)

Cases where throughput results for the PR are worse than latest master (5% drop or greater), results are shown in **red**.

Note that these results are based on a _single_ point-in-time result for each branch. For full results, see [one](https://ddstaging.datadoghq.com/dashboard/fnz-afi-c2i/apm-net-crank) of the [many](https://ddstaging.datadoghq.com/dashboard/c5g-i7d-kcy/ciapp-apm-net-throughput-tests), [many](https://ddstaging.datadoghq.com/dashboard/uxh-m8j-qhi/ciapp-apm-net-throughput-tests-kevin) dashboards!

```mermaid
gantt
    title Throughput Linux x64 (Total requests) 
    dateFormat  X
    axisFormat %s
    section Baseline
    This PR (5208) (11.084M)   : 0, 11084079
    master (11.571M)   : 0, 11570786
    benchmarks/2.9.0 (10.834M)   : 0, 10833884

    section Automatic
    This PR (5208) (7.694M)   : 0, 7693996
    master (7.962M)   : 0, 7962078
    benchmarks/2.9.0 (7.842M)   : 0, 7841734

    section Trace stats
    This PR (5208) (7.869M)   : 0, 7869425
    master (8.204M)   : 0, 8204347

    section Manual
    This PR (5208) (9.622M)   : 0, 9621578
    master (9.982M)   : 0, 9981622

    section Manual + Automatic
    This PR (5208) (6.827M)   : crit ,0, 6827472
    master (7.581M)   : 0, 7581166

    section Version Conflict
    This PR (5208) (6.714M)   : 0, 6714475
    master (6.897M)   : 0, 6896652

```
```mermaid
gantt
    title Throughput Linux arm64 (Total requests) 
    dateFormat  X
    axisFormat %s
    section Baseline
    This PR (5208) (9.650M)   : 0, 9649920
    master (9.383M)   : 0, 9382619
    benchmarks/2.9.0 (9.534M)   : 0, 9534497

    section Automatic
    This PR (5208) (6.584M)   : 0, 6583588
    master (6.560M)   : 0, 6559940

    section Trace stats
    This PR (5208) (6.898M)   : 0, 6898061
    master (6.777M)   : 0, 6776907

    section Manual
    This PR (5208) (8.425M)   : 0, 8425244
    master (8.063M)   : 0, 8063414

    section Manual + Automatic
    This PR (5208) (6.243M)   : 0, 6242955
    master (6.225M)   : 0, 6224526

    section Version Conflict
    This PR (5208) (5.816M)   : 0, 5815993
    master (5.691M)   : 0, 5691136

```
```mermaid
gantt
    title Throughput Windows x64 (Total requests) 
    dateFormat  X
    axisFormat %s
    section Baseline
    This PR (5208) (10.267M)   : 0, 10266695
    master (10.356M)   : 0, 10356131
    benchmarks/2.9.0 (10.178M)   : 0, 10178311

    section Automatic
    This PR (5208) (7.174M)   : 0, 7174374
    master (7.237M)   : 0, 7237455
    benchmarks/2.9.0 (7.446M)   : 0, 7445613

    section Trace stats
    This PR (5208) (7.637M)   : 0, 7636641
    master (7.588M)   : 0, 7588393

    section Manual
    This PR (5208) (8.870M)   : crit ,0, 8870069
    master (9.463M)   : 0, 9463047

    section Manual + Automatic
    This PR (5208) (6.933M)   : 0, 6933309
    master (7.056M)   : 0, 7055849

    section Version Conflict
    This PR (5208) (6.241M)   : 0, 6240626
    master (6.295M)   : 0, 6294767

```
```mermaid
gantt
    title Throughput Linux x64 (ASM) (Total requests) 
    dateFormat  X
    axisFormat %s
    section Baseline
    master (7.316M)   : 0, 7315919
    benchmarks/2.9.0 (7.752M)   : 0, 7751668

    section No attack
    master (1.832M)   : 0, 1832328
    benchmarks/2.9.0 (3.208M)   : 0, 3207733

    section Attack
    master (1.458M)   : 0, 1458489
    benchmarks/2.9.0 (2.458M)   : 0, 2457864

    section Blocking
    master (3.125M)   : 0, 3124924

    section IAST default
    master (6.348M)   : 0, 6348424

    section IAST full
    master (5.463M)   : 0, 5462736

    section Base vuln
    master (0.961M)   : 0, 960884

    section IAST vuln
    master (0.853M)   : 0, 853267

```

File: tracer/src/Datadog.Trace.Tools.dd_dotnet/Checks/Resources.cs (hunk 2 above)
Comment (andrewlock) [reply tracer/src/Datadog.Trace.Tools.dd_dotnet/Checks/Resources.cs:60]: Sure, fair enough

**Review by zacharycmontoya — APPROVED** [review]

**Pull request merged**